### PR TITLE
Remove join on top of lookup-join when only one side is retrieved 

### PR DESCRIPTION
**Pull request by mkleen**
## Summary of the changes / Why this improves CrateDB

This is a follow-up for the lookup-join optimization. The Hash/Nestedloop Join operator can be dropped where there is a lookup-join optimization in place and the operators on top of the join reference only the larger table of the join. This happens as part of the column-pruning logic in the Join Operators because the referenced outputs are needed to make this optimization possible.

Assume the following query:

```
explain select count(t1.a) from doc.t1 join doc.t2 on t1.id = t2.id;
```

```
HashAggregate[count(a)] 
 └ HashJoin[(id = id)]
   ├ MultiPhase
   │  └ Collect[doc.t1 | [a, id] | (id = ANY((doc.t2)))]
   │  └ Collect[doc.t2 | [id] | true]
   └ Collect[doc.t2 | [id] | true]
```
becomes:

```
 HashAggregate[count(a)]                               
   └ MultiPhase                                   
     └ Collect[doc.t1 | [a, id] | (id = ANY((doc.t2)))] 
     └ Collect[doc.t2 | [id] | true]            

```

```
Q: select count(t1.a) from doc.t1 join doc.t2 on t1.id = t2.id
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       45.073 ±   27.470 |     40.679 |     41.745 |     42.544 |    315.736 |
|   V2    |       31.856 ±   26.484 |     27.489 |     28.419 |     29.175 |    292.208 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  34.36%                           -  37.99%
There is a 99.93% probability that the observed difference is not random, and the best estimate of that difference is 34.36%
The test has statistical significance
```

Resolves https://github.com/crate/crate/issues/15003

## Checklist

 - [x] Added an entry in the latest `docs/appendices/release-notes/<x.y.0>.rst` for user facing changes
 - [x] Updated documentation & `sql_features` table for user facing changes
 - [x] Touched code is covered by tests
 - [x] [CLA](https://crate.io/community/contribute/cla/) is signed
 - [x] This does not contain breaking changes, or if it does:
    - It is released within a major release
    - It is recorded in the latest `docs/appendices/release-notes/<x.y.0>.rst`
    - It was marked as deprecated in an earlier release if possible
    - You've thought about the consequences and other components are adapted
      (E.g. AdminUI)

### CODE REVIEW

**Comment by mkleen** [comment]
@mfussenegger Could you give me some early feedback ? It works but the performance boost is nothing special. Shall we get this in ?

**Comment by mfussenegger** [comment]
I'd say 20-30% looks promising enough, so I'd continue.

How does this behave on queries with more data? 5ms is already fairly fast, would be interested seeing the impact on queries that are in the 100ms+ range without it.

**Comment by mkleen** [comment]
> I'd say 20-30% looks promising enough, so I'd continue.
> 
> How does this behave on queries with more data? 5ms is already fairly fast, would be interested seeing the impact on queries that are in the 100ms+ range without it.

100 times more data lead to the following benchmark:

```
Q: select count(t1.a) from doc.t1 join doc.t2 on t1.id = t2.id
C: 1
| Version |         Mean ±    Stdev |        Min |     Median |         Q3 |        Max |
|   V1    |       45.073 ±   27.470 |     40.679 |     41.745 |     42.544 |    315.736 |
|   V2    |       31.856 ±   26.484 |     27.489 |     28.419 |     29.175 |    292.208 |
├---------┴-------------------------┴------------┴------------┴------------┴------------┘
|               -  34.36%                           -  37.99%
There is a 99.93% probability that the observed difference is not random, and the best estimate of that difference is 34.36%
The test has statistical significance
```

File: server/src/main/java/io/crate/planner/operators/AbstractJoinPlan.java [hunk 1]
```diff
@@ -48,15 +48,21 @@ public abstract class AbstractJoinPlan implements LogicalPlan {
     @Nullable
     protected final Symbol joinCondition;
     protected final JoinType joinType;
+    protected final boolean lhsIsLookup;
+    protected final boolean rhsIsLookup;
```
Comment (mkleen) [inline server/src/main/java/io/crate/planner/operators/AbstractJoinPlan.java:52]: An alternative could be to introduce a Lookup-Join-Operator and then check for the type of the sub-plans. However, this approach is simpler and cheaper in terms of resources.

File: server/src/main/java/io/crate/planner/operators/HashJoin.java [hunk 2]
```diff
@@ -180,18 +190,30 @@ public LogicalPlan pruneOutputsExcept(SequencedCollection<Symbol> outputsToKeep)
             SymbolVisitors.intersection(outputToKeep, lhs.outputs(), lhsToKeep::add);
             SymbolVisitors.intersection(outputToKeep, rhs.outputs(), rhsToKeep::add);
         }
-        SymbolVisitors.intersection(joinCondition, lhs.outputs(), lhsToKeep::add);
-        SymbolVisitors.intersection(joinCondition, rhs.outputs(), rhsToKeep::add);
-        LogicalPlan newLhs = lhs.pruneOutputsExcept(lhsToKeep);
-        LogicalPlan newRhs = rhs.pruneOutputsExcept(rhsToKeep);
-        if (newLhs == lhs && newRhs == rhs) {
-            return this;
+        // If the outputs come only from one side and we have a lookup-join in place
```
Comment (matriv) [inline server/src/main/java/io/crate/planner/operators/HashJoin.java:193]: Maybe a bit better phrasing:
```
If there a lookup-join in place, and the outputs belong only to the lookup side, we can drop the join and return only the lookup-side
```

File: server/src/main/java/io/crate/planner/operators/AbstractJoinPlan.java (hunk 1 above)
Comment (matriv) [reply server/src/main/java/io/crate/planner/operators/AbstractJoinPlan.java:52]: I'd say to keep it as is and consider introducing a Lookup-Join if in the future we add more functionality (further optimisations).

**Review by matriv — COMMENTED** [review]
Look good! thx!
Before approving I'd like to ask if it's possible to have a benchmark with nested loops as well. Because of the issue with session settings (disable hashjoins), maybe you can run it using a branch where you hard-code to disable hash joins.

File: server/src/main/java/io/crate/planner/operators/NestedLoopJoin.java [hunk 3]
```diff
@@ -225,9 +230,20 @@ public LogicalPlan pruneOutputsExcept(SequencedCollection<Symbol> outputsToKeep)
...
+
         if (joinCondition != null) {
-            SymbolVisitors.intersection(joinCondition, lhs.outputs(), lhsToKeep::add);
-            SymbolVisitors.intersection(joinCondition, rhs.outputs(), rhsToKeep::add);
+            // If the outputs come only from one side and we have a lookup-join in place
+            // we can drop the join and return only the lookup-side
+            if (lhsToKeep.isEmpty() && rhsIsLookup) {
+                SymbolVisitors.intersection(joinCondition, rhs.outputs(), rhsToKeep::add);
+                return rhs.pruneOutputsExcept(rhsToKeep);
+            } else if (rhsToKeep.isEmpty() && lhsIsLookup) {
```
Comment (mfussenegger) [inline server/src/main/java/io/crate/planner/operators/NestedLoopJoin.java:240]: If lhsIsLookup/rhsIsLookup cannot be both true at the same time, would it make sense to have a `LookupJoin` enum with `LEFT` / `RIGHT` / `NONE` ?

File: server/src/test/java/io/crate/integrationtests/LookupJoinIntegrationTest.java [hunk 4]
```diff
@@ -44,18 +44,64 @@ public void test_lookup_join_with_two_collect_operators() throws Exception {
...
+            "  └ Collect[doc.t2 | [id] | true]"
+        );
+        execute(query);
+        assertThat(response).hasRowCount(100);
+    }
+
+    @UseRandomizedOptimizerRules(0)
+    @UseHashJoins(1)
+    @Test
+    public void test_drop_hashjoin_join_on_top_of_lookup_join() throws Exception {
```
Comment (mfussenegger) [inline server/src/test/java/io/crate/integrationtests/LookupJoinIntegrationTest.java:87]: Maybe worth merging the test cases to share the setup, to reduce the costs of these test cases?

Comment (mkleen) [reply server/src/test/java/io/crate/integrationtests/LookupJoinIntegrationTest.java:87]: We could run on a single test and rely on the randomization.

**Review by matriv — APPROVED** [review]
thx!

**Review by mfussenegger — APPROVED** [review]

**Pull request merged**